osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs

```diff
@@ -0,0 +1,205 @@
+using osu.Game.Rulesets.Osu.UI;
+using osu.Game.Screens.Play;
+using osu.Game.Skinning;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Rulesets.Osu.Skinning.Legacy
+{
+    public class LegacyCursorParticles : CompositeDrawable, IKeyBindingHandler<OsuAction>, IRequireHighFrequencyMousePosition
+    {
+        private const int particle_lifetime_min = 300;
+        private const int particle_lifetime_max = 1000;
+        private const float particle_gravity = 240;
+
+        public bool Active => breakSpewer?.Active.Value == true || kiaiSpewer?.Active.Value == true;
+
+        private Vector2 cursorVelocity;
+        private ParticleSpewer breakSpewer;
+        private ParticleSpewer kiaiSpewer;
+
+        [Resolved(canBeNull: true)]
+        private Player player { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private OsuPlayfield osuPlayfield { get; set; }
+
+        [BackgroundDependencyLoader]
+        private void load(ISkinSource skin, OsuColour colours)
+        {
+            var texture = skin.GetTexture("star2");
+            var starBreakAdditive = skin.GetConfig<OsuSkinColour, Color4>(OsuSkinColour.StarBreakAdditive)?.Value ?? new Color4(255, 182, 193, 255);
+
+            if (texture != null)
+            {
+                // stable "magic ratio". see OsuPlayfieldAdjustmentContainer for full explanation.
+                texture.ScaleAdjust *= 1.6f;
+            }
+
+            RelativeSizeAxes = Axes.Both;
+            Anchor = Anchor.Centre;
+            InternalChildren = new[]
+            {
+                breakSpewer = new ParticleSpewer(texture, 20, particle_lifetime_max)
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Size = Vector2.One,
+                    Colour = starBreakAdditive,
+                    ParticleGravity = particle_gravity,
+                    CreateParticle = createBreakParticle,
+                },
+                kiaiSpewer = new ParticleSpewer(texture, 60, particle_lifetime_max)
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Size = Vector2.One,
+                    Colour = starBreakAdditive,
+                    ParticleGravity = particle_gravity,
+                    CreateParticle = createParticle,
+                },
+            };
+
+            if (player != null)
+            {
+                breakSpewer.Active.BindTarget = player.IsBreakTime;
+            }
+        }
+
+        protected override void Update()
+        {
+            if (osuPlayfield == null) return;
+
+            // find active kiai slider or spinner.
+            var kiaiHitObject = osuPlayfield.HitObjectContainer.AliveObjects.FirstOrDefault(h =>
+                h.HitObject.Kiai &&
+                (
+                    (h is DrawableSlider slider && slider.Tracking.Value) ||
+                    (h is DrawableSpinner spinner && spinner.RotationTracker.Tracking)
+                )
+            );
```

Comment:
When it comes to any allocs here, they will be due to underlying LINQ iterator usage. I don't necessarily thing anything smart can be done regarding them in this pull, it should probably be fixed holistically separately (because other places also use `AliveObjects` in similar ways).

If `OrderedDictionary` wasn't such a terrible collection type it could have been used in `PooledDrawableWithLifetimeContainer`, but alas...

Reply:
Yeah it might be easiest to leave it as is for now, I tested it further and the performance hit of this seems to be smaller than I originally thought (at least on my desktop PC).

I think having something similar to `ApplyToDrawableHitObject` that mods use would be ideal as I'd only need to bind the `Tracking` bindable when a hitobject enters the playfield.

Reply:
I've pushed a solution to avoid this enumeration outside of kiai mode, which I think is an improvement (even though `AliveObjects` is used in other places, it *is* expensive and should be avoided for always-on cases like this where possible).

@bdach Can you take a look and make sure you're okay with the change?

Reply:
Yep, all good with me.